Make item crafting options include recipes from nearby books

wapcaplet opened the pull request:
#### Summary
```SUMMARY: Bugfixes "Make item description consistent with crafting UI for recipe list"```

#### Purpose of change

Resolves #37643


#### Describe the solution

Instead of showing only memorized recipes at the bottom of the item description panel, show all available recipes (including those from nearby books), so the "You could use it to craft: ..." message is consistent with the recipes found in the main crafting UI for that item.

If no recipes are available (in memorized recipes or nearby books), include this at the bottom of the item description: "You know of nothing you could craft with it."


#### Describe alternatives you've considered

Considered preserving the original behavior in the item description of only showing learned/memorized recipes, but was persuaded by other players/developers that this was bad and should be changed.


#### Testing

I tested a variety of ad-hoc use cases, but here's a specific example:

- Create a new world and blank 0-skill character (survivor scenario)
- Use the debugging menu to spawn in:
  - "What's a Transistor?"
  - 1 heating element
  - 1 amplifier circuit
- Read the book or use debug to get Electronics skill to level 2

Then compare:

- With "What's a Transistor?" nearby:
  - Inspect heating element: You could use it to craft: hotplate and coffeemaker
  - Go to crafting menu and `/` to search, `c:heating element`: Matches hotplate and coffemaker
  - Inspect amplifier circuit, and compare recipes with crafting menu: All 7 recipes match
- Without "What's a Transistor?" nearby:
  - amplifier circuit only has "noise emitter (off)" in both screens
  - heating element has no recipes available in either screen

#### Additional context

Screenshots below are cut/pasted to show item description and crafting UI for comparison.

With "What's a Transistor?" nearby, the heating element has two recipes available in both the item description and crafting menu:

![image](https://user-images.githubusercontent.com/238395/73993278-f8289380-490e-11ea-9104-e7de0aa2187f.png)

Similarly, the amplifier circuit item description and crafting recipes are matching:

![image](https://user-images.githubusercontent.com/238395/73993315-05de1900-490f-11ea-9de5-ebfa85947848.png)

With the book too far away, the only recipe I get for the amplifier circuit is the level-0-auto-learned "noise emitter (off)":

![image](https://user-images.githubusercontent.com/238395/73993352-2dcd7c80-490f-11ea-9b64-faeba2edd651.png)

And for the heating element, without "What's a Transistor?", there are no known recipes:

![image](https://user-images.githubusercontent.com/238395/73993536-a6343d80-490f-11ea-8249-793084a0404a.png)

wapcaplet commented:
I see this change caused a couple failing test cases in `tests/iteminfo_test.cpp`.

Marking this PR as WIP until I can fix those (I've been looking for an excuse to get into the tests anyway).

wapcaplet commented:
A misunderstanding on my part regarding the `DESCRIPTION_APPLICABLE_RECIPES` flag caused the unit test failure. I have fixed it, and done a little refactoring as well. Full details in the commit message: https://github.com/CleverRaven/Cataclysm-DDA/pull/37661/commits/f45a1e7e15b26c69365a4f22e73b6eeb9e83d293

I'm satisfied with how this works, but realize it's a far-reaching change, affecting probably a majority of item descriptions, and ultimately adds very little useful information. Now that I've done it, I am not so sure that it should be accepted and merged. It's kind of a waste of text.

So I'm submitting it for consideration and discussion, suggestions of alternatives, criticism, outright ridicule etc.

Pros:
- Tells player the reason there are no crafts available (lack of knowledge)
- Suggests itself as a placeholder, a kind of implicit "This is where you would see craftables if you knew any"

Cons:
- Redundant and tedious - many (if not most) items will always show this message, forever

codemime commented:
> So I'm submitting it for consideration and discussion, suggestions of alternatives, criticism, outright ridicule etc.

Please see the corresponding issue https://github.com/CleverRaven/Cataclysm-DDA/issues/37643#issuecomment-581693181. In short, I think the description should match the information available through the crafting menu.

wapcaplet commented:
@codemime In commit https://github.com/CleverRaven/Cataclysm-DDA/pull/37661/commits/3228bad2554c1647a0ae78d6bc0826cac6cf15de I tried to use `get_available_recipes` as we discussed, passing `crafting_inv = g->u.crafting_inventory()` but I must have done something horribly wrong because now every time I try to use the main crafting menu (which needs item descriptions) or inspect an item that has one or more possible recipes, the game simply crashes with a segmentation fault. The relevant part of the crash log (before getting into low-level string stuff) is all about item info:

```
    /usr/include/c++/7/bits/hashtable.h:1439
    Item_factory::find_template(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
    …/src/item_factory.cpp:1216
    item::item(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, time_point, item::default_charges_tag)
    ??:?
    recipe::create_result() const
    …/src/recipe.cpp:403
    recipe::get_component_filter(recipe_filter_flags) const
    …/src/recipe.cpp:562
    item::final_info(std::vector<iteminfo, std::allocator<iteminfo> >&, iteminfo_query const*, int, bool) const
    ??:?
    item::info[abi:cxx11](std::vector<iteminfo, std::allocator<iteminfo> >&, iteminfo_query const*, int) const
```

I  can post the full log if it'll help, but I'm betting you could look at my code and immediately know what the problem is. It's been something like 20 years since I did any C++ so my syntax may be awkward...

Also, let's just take a moment to appreciate that commit id: `3228bad`.

codemime commented on src/item.cpp:
```diff
@@ -3295,39 +3295,48 @@ void item::final_info( std::vector<iteminfo> &info, const iteminfo_query *parts,
+        // FIXME: Somehow the following lines are causing a game crash whenever the crafting menu is opened
+        const inventory &crafting_inv = g->u.crafting_inventory();
+        const std::set<const recipe *> &item_recipes = g->u.get_available_recipes( crafting_inv,
+                nullptr ).of_component( tid );
```
This is a common pitfall. `of_component()` returns a reference to the result of `get_available_recipes()` which happens to be a temporary because nothing references it. It gets destroyed soon after being constructed. Just assign it to a variable, and it will be fine.
```c++
const recipe_subset available_recipe_subset = g->u.get_available_recipes( crafting_inv );
const std::set<const recipe *> &item_recipes = available_recipe_subset.of_component( tid );
```

wapcaplet replied:
Many thanks! That got it working. Final commits on the way...

codemime replied:
No worries. I remember the first time I fell into that trap: spent hours debugging.

codemime commented:
>Also, let's just take a moment to appreciate that commit id: 3228bad.

Yeah, I suspect that's Linus making fun of our code )). This just gave me an idea. It would be cool to make a generator of commits with funny ids. Some say, SHA-1 was [cracked](https://www.schneier.com/blog/archives/2005/02/sha1_broken.html), so there can be something better than the brute-force approach for this.

wapcaplet commented:
I've adjusted the PR summary and description to reflect the new nature of this fix, and added screenshots confirming the consistent behavior I am seeing now between the two UIs.

codemime commented on src/item.cpp:
```diff
@@ -3295,39 +3295,47 @@ void item::final_info( std::vector<iteminfo> &info, const iteminfo_query *parts,
+            tid = typeId();
+        } else { // use the contained item
+            tid = contents.front().typeId();
+        }
```
Nit: this assignment would be cleaner with a ternary.

wapcaplet replied:
Done!

wapcaplet commented:
Whoops, hold on - that test case isn't working yet. Back in WIP

wapcaplet commented:
The test cases are in much better shape now, but I'm still getting semi-random failures in a couple of them; for examine sometimes the crowbar will show up with a bunch of higher-level recipes (foldable-light frame, wheelchair wheels, pipe bomb, nail rifle etc.) instead of the expected level-0 "makeshift crowbar" only.

Seems like something is bleeding over from other tests; probably I need to clean up / reset the global user `g->u` - I'm already using `empty_skills()` and `empty_traits()` but maybe it takes more than that...

wapcaplet commented:
All tests are passing now. Assuming this passes build checks, it should be ready to merge.
